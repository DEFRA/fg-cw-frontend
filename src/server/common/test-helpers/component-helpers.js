// The following rules to be enabled when "server" dir is pulled apart
// eslint-disable-next-line
import { load } from "cheerio";
// eslint-disable-next-line
import camelCase from "lodash/camelCase";
import { fileURLToPath } from "node:url";
import nunjucks from "nunjucks";
import path from "path";
import * as filters from "../../../config/nunjucks/filters/filters.js";
import * as globals from "../../../config/nunjucks/globals.js";

const dirname = path.dirname(fileURLToPath(import.meta.url));
const nunjucksTestEnv = nunjucks.configure(
  [
    path.normalize(
      path.resolve(dirname, "../../../../node_modules/govuk-frontend/dist/"),
    ),
    path.normalize(path.resolve(dirname, "../templates")),
    path.normalize(path.resolve(dirname, "../components")),
  ],
  {
    trimBlocks: true,
    lstripBlocks: true,
  },
);

Object.entries(globals).forEach(([name, global]) => {
  nunjucksTestEnv.addGlobal(name, global);
});

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksTestEnv.addFilter(name, filter);
});

/**
 * @param {string} componentName
 * @param {object} params
 * @param {string} [callBlock]
 */
export const renderComponent = (componentName, params, callBlock) => {
  const macroPath = `${componentName}/macro.njk`;
  const macroName = `app${
    componentName.charAt(0).toUpperCase() + camelCase(componentName.slice(1))
  }`;
  const macroParams = JSON.stringify(params, null, 2);
  let macroString = `{%- from "${macroPath}" import ${macroName} -%}`;

  if (callBlock) {
    macroString += `{%- call ${macroName}(${macroParams}) -%}${callBlock}{%- endcall -%}`;
  } else {
    macroString += `{{- ${macroName}(${macroParams}) -}}`;
  }

  return load(nunjucksTestEnv.renderString(macroString, {}));
};
