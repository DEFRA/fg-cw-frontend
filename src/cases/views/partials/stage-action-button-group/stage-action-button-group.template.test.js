import { beforeEach, describe, expect, test } from "vitest";
import { renderComponent } from "../../../../server/common/test-helpers/component-helpers.js";

describe.skip("Stage Action Button Group Component", () => {
  let $buttonGroup;

  const mockParams = {
    buttons: [
      {
        label: "Approve",
        testId: "approve-button",
        nextStage: "approved",
      },
      {
        label: "Reject",
        testId: "reject-button",
        nextStage: "rejected",
      },
    ],
    caseId: "123",
  };

  describe("Component with multiple buttons", () => {
    beforeEach(() => {
      $buttonGroup = renderComponent("stageActionButtonGroup", mockParams);
    });

    test("Should render a button group container", () => {
      expect($buttonGroup(".govuk-button-group").length).toBe(1);
    });

    test("Should render all buttons with correct labels", () => {
      const buttons = $buttonGroup("button");
      expect(buttons.length).toBe(2);
      expect(buttons.eq(0).text().trim()).toBe("Approve");
      expect(buttons.eq(1).text().trim()).toBe("Reject");
    });

    test("Should render buttons with correct form actions", () => {
      const buttons = $buttonGroup("button");
      expect(buttons.length).toBe(2);
      expect(buttons.eq(0).attr("formaction")).toBe("/cases/123");
      expect(buttons.eq(1).attr("formaction")).toBe("/cases/123");
    });

    test("Should render buttons with correct attributes", () => {
      const buttons = $buttonGroup("button");
      expect(buttons.eq(0).attr("data-prevent-double-click")).toBe("true");
      expect(buttons.eq(0).attr("class")).toBe("govuk-button");
      expect(buttons.eq(0).attr("data-module")).toBe("govuk-button");
      expect(buttons.eq(0).attr("data-govuk-button-init")).toBe("");
    });
  });

  describe("Component with no buttons", () => {
    beforeEach(() => {
      $buttonGroup = renderComponent("stageActionButtonGroup", {
        buttons: [],
        caseId: "123",
      });
    });

    test("Should not render any buttons", () => {
      expect($buttonGroup("button").length).toBe(0);
    });
  });
});
