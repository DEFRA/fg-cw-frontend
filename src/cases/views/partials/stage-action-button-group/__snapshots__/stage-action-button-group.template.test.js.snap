// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stage-action-button-group > does not render when no buttons provided 1`] = `
"<div class="govuk-button-group">
</div>
"
`;

exports[`stage-action-button-group > renders multiple buttons 1`] = `
"<div class="govuk-button-group">
    <form method="POST">
      <button
        type="submit"
        data-prevent-double-click="true"
        class="govuk-button"
        data-module="govuk-button"
        data-govuk-button-init=""
        formaction="/cases/123"
      >
        Approve
      </button>
    </form>
    <form method="POST">
      <button
        type="submit"
        data-prevent-double-click="true"
        class="govuk-button"
        data-module="govuk-button"
        data-govuk-button-init=""
        formaction="/cases/123"
      >
        Reject
      </button>
    </form>
</div>
"
`;
